### ISSUE TITLE
Callback function gets NoneType return value

### ISSUE THREAD

**Issue by zhuzzzzzz** [pyepics/pyepics#246]
Hello! 
I encounterd a problem here when I established lots of PV connections (about 2000~ PV connections in 10Hz callback rate) using Python multiprocessing. Sometimes the program get the Nonetype return value when the callback function is called but it should be a number, so my program goes wrong.
My callback function is like this:
![image](https://github.com/pyepics/pyepics/assets/57058943/a9a62b01-bd3d-40d4-ab76-302b2e011dc9)

Python exception is like thie:
![image](https://github.com/pyepics/pyepics/assets/57058943/9ed6b9cb-1956-42ec-bc98-71e3dcc8659f)

**Comment by newville** [comment]
@zhuzzzzzz Hm, sorry for the trouble.  I am not sure I've seen that.  It is hard for me to understand how a connected PV would *sometimes* send a "value" value, and sometimes not.

Can you give any more information on this:

- Is it always the same PV that is failing?
- Does it always fail or sometimes work?
- What is the Epics type for the PV (char, float, enum, etc -- is it an array, etc)?  
- Does it happen when not using multiprocessing?

Sorry, I know that is not a quick answer, but I think I don't have one of those...

**Comment by zhuzzzzzz** [comment]
Sorry, I didn't describe it clearly. Epics type for the PV is float and all PVs are the same in code and only different in name, these PVs are created to self add 1 per .1 second start form 0 to infinity(I just want to test the performance of the code in this way). I don't know whether it is the same PV fails but it seems all PV could fail, because all PVs actually are same and run in the same IOC. 
This problem happens when using multiprocessing(I use multiprocessing is to avoid pyepics to share the same chid, in other words, I use multiprocessing to create different ca connection to the PV) and in a large connections from pyepics to the real PV(around 2000 or sometimes more. when the connection amount is small, all seems OK. ), and I havn't seen this problem happen in previous use.

**Comment by zhuzzzzzz** [comment]
In the callback functions it is supposed to use the previous returned value to minus the current returned value, but in running Python suggests error that the current value is None. So it seems like this: when the value of PV changed and the callback function was called, but pyepics didn't return the value of PV but returned None. And all these happens in a situation that multiprocessing are used and large amouts of PV connections are created to do the same thing

**Comment by newville** [comment]
@zhuzzzzzz 
You said, "... because all PVs actually are same and run in the same IOC".  Do you mean it is exactly the same PV, or that it is the same kind of PV?

I cannot recommend making multiple connections to the same PV or deliberately creating multiple chids to the same PV.   I can believe that would cause problems....

But also, a single (non-multiprocess) process to monitor 2000 PVs does not seem that large to me. 

Just to be clear, in a single (non-multiprocess) process, the CA event loop happens in C/C++ code which calls into the Python callback as needed.  Keeping the Python callback functions small can be very helpful (also note that you cannot do other CA calls within the callback).  

I will often have callbacks only set a "value changed" flag (or maybe write into a dict of "new values") and have a main event loop in Python do 
```
newdata = {}
pvs = {}
def onChanges(pvname=None, value=None, timestamp=None, **kws):
     if pvname is not None and value is not None and timestamp is not None:
           newdata[pvname] = (value, timestamp)

for  pvname in BigList_ofPVS:
    pvs[pvname] = epics.get_pv(pvname, callback=onChanges)

time.sleep(0.25) # give some time for connections to happen

while True:
      ca.poll(0.0005, 0.5)    # < - allow even callbacks to run 
      new_pvnames = list(newdata.keys())
      for pvname in new_pvnames:
         val, ts = newdata.pop(pvname)
         # do work with these new values
```

**Comment by zhuzzzzzz** [comment]
Thanks very much for help, I still have some questions here.

It is the same kind of PV, but there also are many connections creating multiple chids to the same PV. I wonder what problem may occur under this situation?

`def onChanges(pvname=None, value=None, timestamp=None, **kws):
     if pvname is not None and value is not None and timestamp is not None:
           newdata[pvname] = (value, timestamp)`
In your code here you add a line " if pvname is not None and value is not None and timestamp is not None: ", I wonder whether this line is necessary? In other words, From my point of view, it is the PV value changed, so then the callback functions are called, it should automatically get the changed values as input args of the func onChanges, isn't it? So why here to judge whether input args is None? Are there other cases that callback function was called but callback functions get NoneType input value？

`ca.poll(0.0005, 0.5)    # < - allow even callbacks to run `
Here you use ca.poll, I wonder whether it is necessary, as I have bind callback functions to the PV, where does these funcitons are called and run? Is in the main thread of Python or in the sub-thread of the main thread or in the C/C++ code? And you set the 0.0005 and 0.5, how do I determine a proper time to set when calling ca.poll? I have read the doc it says that ca.poll will call pend_event and pend_io, what mechanisms are under these two function?

**Comment by newville** [comment]
@zhuzzzzzz 

> It is the same kind of PV, but there also are many connections creating multiple chids to the same PV. I wonder what problem may occur under this situation?

Many connections to the same chid can potentially cause problems for the IOC / client script connections.   But also, why do you need multiple connections?  

In fact, in Pyepics we sort of work to "cache connection information" and keep global (process- or thread-wide) connections and reuse those.  The convenience function  `epics.PV.get_pv()` will either create a new connection and set-up all the callbacks, or use an existing connection (and maybe add more Python callbacks to that).  This is very helpful in, say, long-running GUI code where you might want information about a particular PV reported many times.   You don't need a new connection to the IOC, you just need to get the same value at a few different places in your code.

>def onChanges(pvname=None, value=None, timestamp=None, **kws): if pvname is not None and value is not None and timestamp is not None: newdata[pvname] = (value, timestamp)
In your code here you add a line " if pvname is not None and value is not None and timestamp is not None: ", I wonder whether this line is necessary?

Right, I am not sure if that is necessary.  If the callback is called without pvname, value, or timestamp (for the TIME_*** variants of the native PVs that are used by default), the handler for the "monitor events" clearly got messed up.  
I would interpret any of these values as "data sent from the IOC got corrupted somehow", but I am not sure of that.
It might be worth reporting that instead of just ignoring it.  

> ca.poll(0.0005, 0.5)    # < - allow even callbacks to run 
Here you use ca.poll, I wonder whether it is necessary, as I have bind callback functions to the PV, 


Well, you do have to wait for events to happen. `ca.poll(0.0005, 0.5)` will wait for at 0.0005 seconds, and possibly (but very rarely, in my experience) up to 0.5 seconds for CA events and i/o to finish.   Using that or `time.sleep` (perhaps `time.sleep(0.001)` is sort of necessary to say "suspend Python and allow i/o from CA to happen".    It is also important to wait a short time after creating PVs to allow the connections to happen in the background CA threads, or to assume that for the next fraction of a second, some of the PVs will not report as "connected".

> where does these funcitons are called and run? Is in the main thread of Python or in the sub-thread of the main thread or in the C/C++ code? 

Main Python thread.   I should say that I very rarely use CA in Python Threads and probably honestly never with Python multiprocessing. 

>> And you set the 0.0005 and 0.5, how do I determine a proper time to set when calling ca.poll? I have read the doc it says that ca.poll will call pend_event and pend_io, what mechanisms are under these two function?

For `ca.poll`, the 0.0005 used is "guarantee wait for that long" so that the CA thread can listen for events (new values on subscribed PVs), and the 0.5 is "wait up to that long" for CA to finish its I/O. With the way pyepics uses CA (with PREEMPTIVE_CALLBACKS), having to wait for I/O is only important for large arrays of data. 

So, basically, using a sleep, say between `time.sleep(0.0005)` or `time.sleep(0.005)` will be fine: suspend Python to allow the background CA thread to handle events.  You'll notice that all the callbacks you define will run **during that sleep**.   That does mean you really want to limit the amount of processing done in the callback... The `onChanges` callback just basically records the new values, and then the main thread does any Python processing with those new values.   If that data-processing needs to take a long time, that's where I would recommend using a sub-process, or maybe writing to a database and trigger some completely separate process to handle that task.
